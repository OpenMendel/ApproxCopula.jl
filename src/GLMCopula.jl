module GLMCopula
using Convex, LinearAlgebra, MathProgBase, Reexport, GLM, Distributions, StatsFuns, Statistics, StatsBase, ToeplitzMatrices
using LinearAlgebra: BlasReal, copytri!
using SpecialFunctions
@reexport using Ipopt
@reexport using NLopt

export fit!, fit2!, update_Σ_jensen!, init_β!, initialize_model!, loglikelihood!, standardize_res!, std_res_differential!
export update_res!, update_Σ!
export update_∇Σ!, update_HΣ! # update gradient and hessian of variance components
export glm_regress_jl, glm_regress_model, glm_score_statistic!  # these are to initialize our model
export component_loglikelihood, glm_gradient, hessian_glm
export GLMCopulaVCObs, GLMCopulaVCModel

mutable struct GLMCopulaVCObs{T <: BlasReal, D, Link}
    # data
    y::Vector{T}
    X::Matrix{T}
    V::Vector{Matrix{T}}
    # working arrays
    ∇β::Vector{T}   # gradient wrt β
    ∇μβ::Matrix{T}
    ∇σ2β::Matrix{T}
    ∇resβ::Matrix{T}# residual gradient matrix d/dβ_p res_ij (each observation has a gradient of residual is px1)
    ∇τ::Vector{T}   # gradient wrt τ
    ∇Σ::Vector{T}   # gradient wrt σ2
    Hβ::Matrix{T}   # Hessian wrt β
    HΣ::Matrix{T}   # Hessian wrt variance components Σ
    Hτ::Matrix{T}   # Hessian wrt τ
    res::Vector{T}  # residual vector res_i
    t::Vector{T}    # t[k] = tr(V_i[k]) / 2
    q::Vector{T}    # q[k] = res_i' * V_i[k] * res_i / 2
    xtx::Matrix{T}  # Xi'Xi
    storage_n::Vector{T}
    m1::Vector{T}
    m2::Vector{T}
    storage_p1::Vector{T}
    storage_p2::Vector{T}
    storage_np::Matrix{T}
    storage_pp::Matrix{T}
    added_term_numerator::Matrix{T}
    added_term2::Matrix{T}
    η::Vector{T}    # η = Xβ systematic component
    μ::Vector{T}    # μ(β) = ginv(Xβ) # inverse link of the systematic component
    varμ::Vector{T} # v(μ_i) # variance as a function of the mean
    dμ::Vector{T}   # derivative of μ
    d::D            # distribution()
    link::Link      # link function ()
    wt::Vector{T}   # weights wt for GLM.jl
    w1::Vector{T}   # working weights in the gradient = dμ/v(μ)
    w2::Vector{T}   # working weights in the information matrix = dμ^2/v(μ)
end

function GLMCopulaVCObs(
    y::Vector{T},
    X::Matrix{T},
    V::Vector{Matrix{T}},
    d::D,
    link::Link) where {T <: BlasReal, D, Link}
    n, p, m = size(X, 1), size(X, 2), length(V)
    @assert length(y) == n "length(y) should be equal to size(X, 1)"
    # working arrays
    ∇β  = Vector{T}(undef, p)
    ∇μβ = Matrix{T}(undef, n, p)
    ∇σ2β = Matrix{T}(undef, n, p)
    ∇resβ  = Matrix{T}(undef, n, p)
    ∇τ  = Vector{T}(undef, 1)
    ∇Σ  = Vector{T}(undef, m)
    Hβ  = Matrix{T}(undef, p, p)
    HΣ  = Matrix{T}(undef, m, m)
    Hτ  = Matrix{T}(undef, 1, 1)
    res = Vector{T}(undef, n)
    t   = [tr(V[k])/2 for k in 1:m]
    q   = Vector{T}(undef, m)
    xtx = transpose(X) * X
    storage_n = Vector{T}(undef, n)
    m1        = Vector{T}(undef, m)
    m2        = Vector{T}(undef, m)
    storage_p1 = Vector{T}(undef, p)
    storage_p2 = Vector{T}(undef, p)
    storage_np = Matrix{T}(undef, n, p)
    storage_pp = Matrix{T}(undef, p, p)
    added_term_numerator = Matrix{T}(undef, n, p)
    added_term2 = Matrix{T}(undef, p, p)
    η = Vector{T}(undef, n)
    μ = Vector{T}(undef, n)
    varμ = Vector{T}(undef, n)
    dμ = Vector{T}(undef, n)
    wt = Vector{T}(undef, n)
    fill!(wt, one(T))
    w1 = Vector{T}(undef, n)
    w2 = Vector{T}(undef, n)
    # constructor
    GLMCopulaVCObs{T, D, Link}(y, X, V, ∇β, ∇μβ, ∇σ2β, ∇resβ, ∇τ, ∇Σ, Hβ, HΣ,
        Hτ, res, t, q, xtx, storage_n, m1, m2, storage_p1, storage_p2, storage_np, storage_pp, added_term_numerator, added_term2, η, μ, varμ, dμ, d, link, wt, w1, w2)
end

"""
GLMCopulaVCModel
GLMCopulaVCModel(gcs)
Gaussian copula variance component model, which contains a vector of
`GLMCopulaVCObs` as data, model parameters, and working arrays.
"""
struct GLMCopulaVCModel{T <: BlasReal, D, Link} <: MathProgBase.AbstractNLPEvaluator
    # data
    data::Vector{GLMCopulaVCObs{T, D, Link}}
    Ytotal::T
    ntotal::Int     # total number of singleton observations
    p::Int          # number of mean parameters in linear regression
    m::Int          # number of variance components
    # parameters
    β::Vector{T}    # p-vector of mean regression coefficients
    τ::Vector{T}    # inverse of linear regression variance parameter
    Σ::Vector{T}    # m-vector: [σ12, ..., σm2]
    θ::Vector{T}
    # working arrays
    ∇β::Vector{T}   # gradient from all observations
    ∇τ::Vector{T}
    ∇Σ::Vector{T}
    ∇θ::Vector{T}   # overall gradient for beta and variance components vector Σ
    XtX::Matrix{T}  # X'X = sum_i Xi'Xi
    Hβ::Matrix{T}    # Hessian from all observations
    HΣ::Matrix{T}
    Hτ::Matrix{T}
    Ainv::Matrix{T}
    Aevec::Matrix{T}
    M::Matrix{T}
    vcov::Matrix{T}
    ψ::Vector{T}
    TR::Matrix{T}         # n-by-m matrix with tik = tr(Vi[k]) / 2
    QF::Matrix{T}         # n-by-m matrix with qik = res_i' Vi[k] res_i
    storage_n::Vector{T}
    storage_m::Vector{T}
    storage_Σ::Vector{T}
    d::Vector{D}
    link::Vector{Link}
end

function GLMCopulaVCModel(gcs::Vector{GLMCopulaVCObs{T, D, Link}}) where {T <: BlasReal, D, Link}
    n, p, m = length(gcs), size(gcs[1].X, 2), length(gcs[1].V)
    β       = Vector{T}(undef, p)
    τ       = [1.0]
    Σ       = Vector{T}(undef, m)
    θ       = Vector{T}(undef, m + p)
    ∇β      = Vector{T}(undef, p)
    ∇τ      = Vector{T}(undef, 1)
    ∇Σ      = Vector{T}(undef, m)
    ∇θ      = Vector{T}(undef, m + p)
    XtX     = zeros(T, p, p) # sum_i xi'xi
    Hβ      = Matrix{T}(undef, p, p)
    HΣ      = Matrix{T}(undef, m, m)
    Hτ      = Matrix{T}(undef, 1, 1)
    Ainv    = zeros(T, p + m, p + m)
    Aevec   = zeros(T, p + m, p + m)
    M       = zeros(T, p + m, p + m)
    vcov    = zeros(T, p + m, p + m)
    ψ       = Vector{T}(undef, p + m)
    TR      = Matrix{T}(undef, n, m) # collect trace terms
    Ytotal  = 0.0
    ntotal  = 0.0
    d       = Vector{D}(undef, n)
    link    = Vector{Link}(undef, n)
    for i in eachindex(gcs)
        ntotal  += length(gcs[i].y)
        Ytotal  += sum(gcs[i].y)
        BLAS.axpy!(one(T), gcs[i].xtx, XtX)
        TR[i, :] = gcs[i].t
        d[i] = gcs[i].d
        link[i] = gcs[i].link
    end
    QF        = Matrix{T}(undef, n, m)
    storage_n = Vector{T}(undef, n)
    storage_m = Vector{T}(undef, m)
    storage_Σ = Vector{T}(undef, m)
    GLMCopulaVCModel{T, D, Link}(gcs, Ytotal, ntotal, p, m, β, τ, Σ, θ,
        ∇β, ∇τ, ∇Σ, ∇θ, XtX, Hβ, HΣ, Hτ, Ainv, Aevec, M, vcov, ψ, TR, QF,
        storage_n, storage_m, storage_Σ, d, link)
end

include("parameter_estimation/autoregressive.jl")
include("parameter_estimation/NBCopulaVC.jl")
include("generate_random_deviates/discrete_rand.jl")
include("generate_random_deviates/continuous_rand.jl")
include("generate_random_deviates/multivariate_rand.jl")
include("parameter_estimation/initialize_model.jl")
include("parameter_estimation/splitting_loglikelihood.jl")
include("parameter_estimation/gradient_hessian.jl")
include("parameter_estimation/update_sigma_and_residuals.jl")
include("parameter_estimation/fit_ar.jl")
include("parameter_estimation/fit_new.jl") # only initializes using MM-algorithm does joint estimation using newton after
include("parameter_estimation/fit_nb.jl")
include("parameter_estimation/inference_ci.jl")
# include("parameter_estimation/fit_old.jl") # only uses MM-algorithm
end # module    