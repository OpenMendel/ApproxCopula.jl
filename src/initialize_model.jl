"""
initialize!(gcm, Normal)

Initialize the linear regression parameters `β` and `τ=σ0^{-2}` by the least
squares solution.
"""
function initialize_model!(
    gcm::Union{GLMCopulaVCModel{T, D}, GaussianCopulaVCModel{T, D}}) where {T <:BlasReal, D<:Normal}
    # accumulate sufficient statistics X'y
    xty = zeros(T, gcm.p)
    for i in eachindex(gcm.data)
        BLAS.gemv!('T', one(T), gcm.data[i].X, gcm.data[i].y, one(T), xty)
    end
    # least square solution for β s.t gcm.β = inv(cholesky(Symmetric(gcm.XtX)))*xty
    ldiv!(gcm.β, cholesky(Symmetric(gcm.XtX)), xty)
    # accumulate residual sum of squares
    rss = zero(T)
    for i in eachindex(gcm.data)
        update_res!(gcm.data[i], gcm.β)
        rss += abs2(norm(gcm.data[i].res))
    end
    gcm.τ[1] = gcm.ntotal / rss
    gcm.β
end

"""
initialize_model!(gcm)

Initialize the linear regression parameters `β` by the weighted least
squares solution.
"""
function initialize_model!(
    gcm::Union{GLMCopulaVCModel{T, D}, GaussianCopulaVCModel{T, D}}) where {T <: BlasReal, D}
    glm_β = glm_regress_model(gcm)
    copyto!(gcm.β, glm_β)
    fill!(gcm.τ, 1.0)
    nothing
end

"""
glm_regress_model(gcm)

Initialize beta for glm model
"""
function glm_regress_model(gcm::Union{GLMCopulaVCModel{T, D}, GaussianCopulaVCModel{T, D}}) where {T <:BlasReal, D}
  (n, p) = gcm.ntotal, gcm.p
   beta = zeros(p)
   ybar = gcm.Ytotal / n
   for iteration = 1:20 # find the intercept by Newton's method
     g1 = GLM.linkinv(gcm.link[1], beta[1]) #  mu
     g2 = GLM.mueta(gcm.link[1], beta[1])  # dmu
     beta[1] =  beta[1] - clamp((g1 - ybar) / g2, -1.0, 1.0)
     if abs(g1 - ybar) < 1e-10
       break
     end
   end
   (obj, old_obj, c) = (0.0, 0.0, 0.0)
   epsilon = 1e-8
   for iteration = 1:100 # scoring algorithm
    fill!(gcm.∇β, 0.0)
    fill!(gcm.Hβ, 0.0)
    gcm = glm_score_statistic(gcm, beta)
    increment = gcm.Hβ \ gcm.∇β
    BLAS.axpy!(1, increment, beta)
    steps = -1
    for step_halve = 0:3 # step halving
      obj = 0.0
      fill!(gcm.∇β, 0.0)
      fill!(gcm.Hβ, 0.0)
           for i in 1:length(gcm.data)
               gc = gcm.data[i]
               x = zeros(p)
              mul!(gc.η, gc.X, beta) # z = X * beta
              update_res!(gc, beta)
              steps = steps + 1
                  for j = 1:length(gcm.data[i].y)
                    c = gc.res[j] * gc.w1[j]
                    copyto!(x, gc.X[j, :])
                    BLAS.axpy!(c, x, gcm.∇β) # score = score + c * x
                    obj = obj + GLMCopula.loglik_obs(gc.d, gc.y[j], gc.μ[j], gc.wt[j], 1)
                   end
           end
      if obj > old_obj
        break
      else
        BLAS.axpy!(-1, increment, beta)
        increment = 0.5 * increment
      end
    end
    println(iteration," ",old_obj," ",obj," ",steps)
    if iteration > 1 && abs(obj - old_obj) < epsilon * (abs(old_obj) + 1.0)
      return beta
    else
      old_obj = obj
    end
   end
    # fill!(gcm.∇β, 0.0)
    # fill!(gcm.Hβ, 0.0)
    println("reach here")
    gcm = glm_score_statistic(gcm, beta)
    increment = gcm.Hβ \ gcm.∇β
    BLAS.axpy!(1, increment, beta)
    return beta
end # function glm_regress

"""
glm_regress_model(gcm)

Initialize beta for glm model
"""
function glm_score_statistic(gc::Union{GLMCopulaVCObs{T, D}, GaussianCopulaVCObs{T, D}},
   β::Vector{T}) where {T<: BlasReal, D}
  (n, p) = size(gc.X)
  x = zeros(p)
  @assert n == length(gc.y)
  @assert p == length(β)
  fill!(gc.∇β, 0.0)
  fill!(gc.Hβ, 0.0)
  mul!(gc.η, gc.X, β) # z = X * beta
  update_res!(gc, β)
  for i = 1:n
    c = gc.res[i] * gc.w1[i]
    copyto!(x, gc.X[i, :])
    BLAS.axpy!(c, x, gc.∇β) # gc.∇β = gc.∇β + r_ij(β) * mueta* x
    BLAS.ger!(gc.w2[i], x, x, gc.Hβ) # gc.Hβ = gc.Hβ + r_ij(β) * x * x'
  end
  return gc
end # function glm_score_statistic

"""
glm_score_statistic(gcm)

Get score
"""
function glm_score_statistic(gcm::Union{GLMCopulaVCModel{T, D}, GaussianCopulaVCModel{T, D}},
   β::Vector) where {T <: BlasReal, D}
  fill!(gcm.∇β, 0.0)
  fill!(gcm.Hβ, 0.0)
    for i in 1:length(gcm.data)
        gcm.data[i] = glm_score_statistic(gcm.data[i], β)
        gcm.∇β .+= gcm.data[i].∇β
        gcm.Hβ .+= gcm.data[i].Hβ
    end
  return gcm
end # function glm_score_statistic
#
# """
# glm_component_score(gc)
#
# get portion of score that is coming from just the glm
# """
# function glm_component_score(gc::Union{GLMCopulaVCObs{T, D}, GaussianCopulaVCObs{T, D}},
#    β::Vector{T}) where {T<: BlasReal, D}
#   (n, p) = size(gc.X)
#   x = zeros(p)
#   @assert n == length(gc.y)
#   @assert p == length(β)
#   fill!(gc.∇β, 0.0)
#   fill!(gc.Hβ, 0.0)
#   mul!(gc.η, gc.X, β) # z = X * beta
#   update_res!(gc, β)
#   for i = 1:n
#     c = gc.res[i] * gc.w1[i]
#     copyto!(x, gc.X[i, :])
#     BLAS.axpy!(c, x, gc.∇β) # gc.∇β = gc.∇β + r_ij(β) * mueta* x
#     BLAS.ger!(gc.w2[i], x, x, gc.Hβ) # gc.Hβ = gc.Hβ + r_ij(β) * x * x'
#   end
# # increment = gc.Hβ \ gc.∇β
# # score_statistic = dot(gc.∇β, increment)
#   return gc
# end # function glm_score_statistic
#
# #  get score from the full model
# function glm_component_score(gcm::GLMCopulaVCModel{T, D},
#    β::Vector) where {T <: BlasReal, D}
#   fill!(gcm.∇β, 0.0)
#   fill!(gcm.Hβ, 0.0)
#     for i in 1:length(gcm.data)
#         gcm.data[i] = glm_component_score(gcm.data[i], β)
#         gcm.∇β .+= gcm.data[i].∇β
#         gcm.Hβ .+= gcm.data[i].Hβ
#     end
#   return gcm
# end # function glm_score_statistic

# function initialize_glm_object()
#   d = Binomial
#   l = canonicallink(d())
#   Random.seed!(12345)
#   xcov = randn(2)
#   X = zeros(200, 2)
#   for i in 1:200
#     X[i, :] .= xcov
#   end
#   y = []
#   for i in 1:length(Y_nsample)
#     push!(y, vec(Y_nsample[i])...)
#     end
#   return fit(GeneralizedLinearModel, X, y, d(), l)
# end