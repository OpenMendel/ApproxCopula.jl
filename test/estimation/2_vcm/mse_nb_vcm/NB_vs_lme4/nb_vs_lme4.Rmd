---
title: "negativebinomial_qc_vs_lme4_glmer"
author: "sarah"
date: "10/26/2021"
output:
  pdf_document: default
  html_document: default
geometry: margin = 1cm
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("lme4")
library("glmmTMB")
```

```{r, echo = FALSE}
data = read.csv("/Users/sarahji/.julia/dev/GLMCopula/test/estimation/2_vcm/mse_nb_vcm/NB_vs_lme4/N10k_ni5_NB.csv", header = TRUE)
```

I simulated a single dataset with sample size 10,000 and cluster size 5 under our negative binomial random intercept GLMM with $\beta \sim$ Uniform(-0.2, 0.2), single variance component Sigma = 0.01 and r = 10. Here I compare the fit of our NB regression with two most popular R packages for fitting Negative Binomial GLMM's with options to estimate "r":
  
  1. lme4: The function glmer.nb estimates r pretty well, but very slowly. To get confidence intervals, it takes a significant amount of additional time, but no inference on r.
  2. glmmTMB: The function glmmTMB estimates r pretty well and a little faster than lme4 but still slow. We can get the CI for all parameters in much less time than lme4.

\newpage


# lme4: glmer.nb
```{r, warning= FALSE}
# Start the clock!
ptm <- proc.time()
m.nb <- glmer.nb(Y ~ 1 + X2 + X3 + (1|group), data = data, verbose = TRUE)
# Stop the clock
proc.time() - ptm
```

### Show estimates
```{r}
m.nb
```

### Show estimated r from lme4: glmer.nb
```{r}
getME(m.nb, "glmer.nb.theta")
```

### Show confidence intervals from lme4: glmer.nb
```{r}
ptm <- proc.time()
confint(m.nb)
# Stop the clock
proc.time() - ptm
```

\newpage

# glmmTMB: glmmTMB

```{r, warning=FALSE}
# Start the clock!
ptm <- proc.time()
m.glmmtmb_nb <- glmmTMB(Y ~ 1 + X2 + X3 + (1|group), data = data, family=nbinom2)
# Stop the clock
proc.time() - ptm
```

### Show estimates
```{r}
m.glmmtmb_nb
```

### Show estimated r from glmmTMB
```{r}
sigma(m.glmmtmb_nb)
```

### Show confidence intervals from glmmTMB
```{r}
ptm <- proc.time()
confint(m.glmmtmb_nb, full = TRUE)
# Stop the clock
proc.time() - ptm
```

