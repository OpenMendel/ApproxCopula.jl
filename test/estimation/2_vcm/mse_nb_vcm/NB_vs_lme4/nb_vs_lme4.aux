\relax 
\providecommand\hyper@newdestlabel[2]{}
\providecommand\HyperFirstAtBeginDocument{\AtBeginDocument}
\HyperFirstAtBeginDocument{\ifx\hyper@anchor\@undefined
\global\let\oldcontentsline\contentsline
\gdef\contentsline#1#2#3#4{\oldcontentsline{#1}{#2}{#3}}
\global\let\oldnewlabel\newlabel
\gdef\newlabel#1#2{\newlabelxx{#1}#2}
\gdef\newlabelxx#1#2#3#4#5#6{\oldnewlabel{#1}{{#2}{#3}}}
\AtEndDocument{\ifx\hyper@anchor\@undefined
\let\contentsline\oldcontentsline
\let\newlabel\oldnewlabel
\fi}
\fi}
\global\let\hyper@last\relax 
\gdef\HyperFirstAtBeginDocument#1{#1}
\providecommand\HyField@AuxAddToFields[1]{}
\providecommand\HyField@AuxAddToCoFields[2]{}
\@writefile{toc}{\contentsline {section}{lme4: glmer.nb}{2}{section*.1}\protected@file@percent }
\newlabel{lme4-glmer.nb}{{}{2}{lme4: glmer.nb}{section*.1}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show estimates}{3}{section*.2}\protected@file@percent }
\newlabel{show-estimates}{{}{3}{Show estimates}{section*.2}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show estimated r from lme4: glmer.nb}{3}{section*.3}\protected@file@percent }
\newlabel{show-estimated-r-from-lme4-glmer.nb}{{}{3}{Show estimated r from lme4: glmer.nb}{section*.3}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show confidence intervals from lme4: glmer.nb}{3}{section*.4}\protected@file@percent }
\newlabel{show-confidence-intervals-from-lme4-glmer.nb}{{}{3}{Show confidence intervals from lme4: glmer.nb}{section*.4}{}}
\@writefile{toc}{\contentsline {section}{glmmTMB: glmmTMB}{4}{section*.5}\protected@file@percent }
\newlabel{glmmtmb-glmmtmb}{{}{4}{glmmTMB: glmmTMB}{section*.5}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show estimates}{4}{section*.6}\protected@file@percent }
\newlabel{show-estimates-1}{{}{4}{Show estimates}{section*.6}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show estimated r from glmmTMB}{4}{section*.7}\protected@file@percent }
\newlabel{show-estimated-r-from-glmmtmb}{{}{4}{Show estimated r from glmmTMB}{section*.7}{}}
\@writefile{toc}{\contentsline {subsubsection}{Show confidence intervals from glmmTMB}{4}{section*.8}\protected@file@percent }
\newlabel{show-confidence-intervals-from-glmmtmb}{{}{4}{Show confidence intervals from glmmTMB}{section*.8}{}}
\gdef \@abspage@last{5}
