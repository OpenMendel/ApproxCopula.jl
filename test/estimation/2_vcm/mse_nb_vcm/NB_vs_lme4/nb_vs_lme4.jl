using GLMCopula, DelimitedFiles, LinearAlgebra, Random, GLM, MixedModels, CategoricalArrays
using Random, Roots, SpecialFunctions
using DataFrames, Statistics
import StatsBase: sem

# simulation 1: generate data under QC model
# simulate samplesize = 10,000 vectors of nb base distribution to save to csv and analyze with R package lme4
p_fixed = 3    # number of fixed effects, including intercept
m = 1    # number of variance components
# true parameter values
βtrue = ones(p_fixed)
Σtrue = [0.1]
rtrue = 10.0

# #simulation parameters
samplesize = 10000
ni = 5 # number of observations per individual

Γ = Σtrue[1] * ones(ni, ni)

d = NegativeBinomial()
link = LogLink()
D = typeof(d)
Link = typeof(link)
T = Float64

gcs = Vector{NBCopulaVCObs{T, D, Link}}(undef, samplesize)

# a = collect(1:samplesize)
# group = [repeat([a[i]], ni) for i in 1:samplesize]
# groupstack = vcat(group...)
# Xstack = []
# Ystack = []
# df = DataFrame(Y = Ystack, X1 = Xstack[:, 1], X2 = Xstack[:, 2], X3 = Xstack[:, 3], group = CategoricalArray(groupstack))
for i in 1:samplesize
    Random.seed!(1234 * i)
    X = [ones(ni) randn(ni, p_fixed - 1)]
    η = X * βtrue
    μ = exp.(η)
    p = rtrue ./ (μ .+ rtrue)
    vecd = Vector{DiscreteUnivariateDistribution}(undef, ni)
    vecd = [NegativeBinomial(rtrue, p[i]) for i in 1:ni]
    nonmixed_multivariate_dist = NonMixedMultivariateDistribution(vecd, Γ)
    # simuate single vector y
    y = Vector{Float64}(undef, ni)
    res = Vector{Float64}(undef, ni)
    Random.seed!(1234 * i)
    rand(nonmixed_multivariate_dist, y, res)
    V = [ones(ni, ni)]
    gcs[i] = NBCopulaVCObs(y, X, V, d, link)
    # push!(Xstack, X)
    # push!(Ystack, y)
end
gcm = NBCopulaVCModel(gcs);
fittime = @elapsed GLMCopula.fit!(gcm, maxBlockIter=100 , tol = 1e-7)
@show fittime
@show gcm.β
@show gcm.Σ
@show gcm.r
@show gcm.θ
@show gcm.∇θ
function get_CI(gcm)
    loglikelihood!(gcm, true, true)
    # @show gcm.θ
    # @show gcm.∇θ
    vcov!(gcm)
    GLMCopula.confint(gcm)
end

# Xstack = [vcat(Xstack...)][1]
# Ystack = [vcat(Ystack...)][1]
# # p = 3
# df = DataFrame(Y = Ystack, X2 = Xstack[:, 2], X3 = Xstack[:, 3], group = string.(groupstack))

# using CSV
# CSV.write(file = "")

# simulation 2: Under GLMM
using GLMCopula, DelimitedFiles, LinearAlgebra, Random, GLM, MixedModels, CategoricalArrays
using Random, Roots, SpecialFunctions
using DataFrames, DelimitedFiles, Statistics
using LinearAlgebra: BlasReal, copytri!

import StatsBase: sem
import GLMCopula.initialize_model!

function __get_distribution(dist::Type{D}, μ) where D <: UnivariateDistribution
    return dist(μ)
end

function __get_distribution(dist::Type{D}, μ, r) where D <: UnivariateDistribution
    return dist(r, μ)
end

function initialize_model!(
    gcm::NBCopulaVCModel{T, D, Link}) where {T <: BlasReal, D, Link}

    # initial guess for r = 1
    fill!(gcm.r, 1)
    fill!(gcm.τ, 1)
    fill!(gcm.β, 0)

    println("initializing variance components using MM-Algorithm")
    fill!(gcm.Σ, 1)
    update_Σ!(gcm)
    nothing
end

p_fixed = 3    # number of fixed effects, including intercept
m = 1    # number of variance components
# true parameter values
Random.seed!(1234)
βtrue = rand(Uniform(-0.2, 0.2), p_fixed)
Σtrue = [0.01]
rtrue = 10.0

# #simulation parameters
samplesize = 10000
ni = 5 # number of observations per individual

d = NegativeBinomial()
link = LogLink()
D = typeof(d)
Link = typeof(link)
T = Float64

gcs = Vector{NBCopulaVCObs{T, D, Link}}(undef, samplesize)

Γ = Σtrue[1] * ones(ni, ni) + 0.00000000000001 * Matrix(I, ni, ni)
a = collect(1:samplesize)
group = [repeat([a[i]], ni) for i in 1:samplesize]
groupstack = vcat(group...)
Xstack = []
Ystack = []
for i in 1:samplesize
    # Random.seed!(1000000000 * t + 10000000 * j + 1000000 * k + i)
    X = [ones(ni) randn(ni, p_fixed - 1)]
    η = X * βtrue
    V = [ones(ni, ni)]
    # generate mvn response
    mvn_d = MvNormal(η, Γ)
    mvn_η = rand(mvn_d)
    μ = GLM.linkinv.(link, mvn_η)
    p_nb = rtrue ./ (μ .+ rtrue)
    y = Float64.(rand.(__get_distribution.(D, p_nb, rtrue)))
    # add to data
    gcs[i] = NBCopulaVCObs(y, X, V, d, link)
    push!(Xstack, X)
    push!(Ystack, y)
end

# form NBCopulaVCModel
gcm = NBCopulaVCModel(gcs);
fittime = @elapsed GLMCopula.fit!(gcm, maxBlockIter=100 , tol = 1e-7)
@show fittime
@show gcm.β
@show gcm.Σ
@show gcm.r
@show gcm.θ
@show gcm.∇θ
function get_CI(gcm)
    loglikelihood!(gcm, true, true)
    # @show gcm.θ
    # @show gcm.∇θ
    vcov!(gcm)
    GLMCopula.confint(gcm)
end
get_CI(gcm)

# form glmm
Xstack = [vcat(Xstack...)][1]
Ystack = [vcat(Ystack...)][1]
# p = 3
df = DataFrame(Y = Ystack, X2 = Xstack[:, 2], X3 = Xstack[:, 3], group = string.(groupstack))

using CSV
CSV.write("N10k_ni5_NB.csv", df)
