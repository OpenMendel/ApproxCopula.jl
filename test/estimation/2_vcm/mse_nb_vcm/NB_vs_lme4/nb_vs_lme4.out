\BOOKMARK [1][-]{section*.1}{\376\377\000l\000m\000e\0004\000:\000\040\000g\000l\000m\000e\000r\000.\000n\000b}{}% 1
\BOOKMARK [2][-]{section*.2}{\376\377\000S\000h\000o\000w\000\040\000e\000s\000t\000i\000m\000a\000t\000e\000s}{section*.1}% 2
\BOOKMARK [3][-]{section*.3}{\376\377\000S\000h\000o\000w\000\040\000e\000s\000t\000i\000m\000a\000t\000e\000d\000\040\000r\000\040\000f\000r\000o\000m\000\040\000l\000m\000e\0004\000:\000\040\000g\000l\000m\000e\000r\000.\000n\000b}{section*.2}% 3
\BOOKMARK [3][-]{section*.4}{\376\377\000S\000h\000o\000w\000\040\000c\000o\000n\000f\000i\000d\000e\000n\000c\000e\000\040\000i\000n\000t\000e\000r\000v\000a\000l\000s\000\040\000f\000r\000o\000m\000\040\000l\000m\000e\0004\000:\000\040\000g\000l\000m\000e\000r\000.\000n\000b}{section*.2}% 4
\BOOKMARK [1][-]{section*.5}{\376\377\000g\000l\000m\000m\000T\000M\000B\000:\000\040\000g\000l\000m\000m\000T\000M\000B}{}% 5
\BOOKMARK [2][-]{section*.6}{\376\377\000S\000h\000o\000w\000\040\000e\000s\000t\000i\000m\000a\000t\000e\000s}{section*.5}% 6
\BOOKMARK [3][-]{section*.7}{\376\377\000S\000h\000o\000w\000\040\000e\000s\000t\000i\000m\000a\000t\000e\000d\000\040\000r\000\040\000f\000r\000o\000m\000\040\000g\000l\000m\000m\000T\000M\000B}{section*.6}% 7
\BOOKMARK [3][-]{section*.8}{\376\377\000S\000h\000o\000w\000\040\000c\000o\000n\000f\000i\000d\000e\000n\000c\000e\000\040\000i\000n\000t\000e\000r\000v\000a\000l\000s\000\040\000f\000r\000o\000m\000\040\000g\000l\000m\000m\000T\000M\000B}{section*.6}% 8
